src/Data/Nat/Primality/Factorisation.agda

```diff
@@ -0,0 +1,218 @@
+  factoriseRec (2+ n) rec {suc (suc k)} 2≤2+n (≤‴-step k<n) k-rough-n with 2 + k ∣? 2+ n
+  ... | no  k∤n = factoriseRec (2+ n) rec {3 + k} 2≤2+n k<n (extend-∤ k-rough-n k∤n)
+  ... | yes k∣n = record
+    { factors = 2 + k ∷ factors res
+    ; isFactorisation = prop
+    ; factorsPrime = roughn∧∣n⇒prime k-rough-n 2≤2+n k∣n ∷ factorsPrime res
+    }
+    where
+      open ≤-Reasoning
+
+      q : ℕ
+      q = quotient k∣n
+
+      -- we know that k < n, so if q is 0 or 1 then q * k < n
+      2≤q : 2 ≤ q
+      2≤q = ≮⇒≥ (λ q<2 → contradiction (_∣_.equality k∣n) (>⇒≢ (prf (≤-pred q<2)))) where
+
+        prf : q ≤ 1 → q * 2+ k < 2 + n
+        prf q≤1 = begin-strict
+          q * 2+ k  ≤⟨ *-monoˡ-≤ (2 + k) q≤1 ⟩
+          2 + k + 0 ≡⟨ +-identityʳ (2 + k) ⟩
+          2 + k     <⟨ ≤‴⇒≤ k<n ⟩
+          2 + n     ∎
+
+      0<q : 0 < q
+      0<q = begin-strict
+        0 <⟨ s≤s z≤n ⟩
+        2 ≤⟨ 2≤q ⟩
+        q ∎
+
+      q<n : q < 2 + n
+      q<n = begin-strict
+        q           <⟨ m<m*n q (2 + k) ⦃ >-nonZero 0<q ⦄ 2≤2+n ⟩
+        q * (2 + k) ≡˘⟨ _∣_.equality k∣n ⟩
+        2 + n       ∎
+
+      q≮2+k : q ≮ 2 + k
+      q≮2+k q<k = k-rough-n 2≤q q<k (quotient∣n k∣n)
+
+      res : PrimeFactorisation q
+      res = rec q q<n {2 + k} 2≤q (≤⇒≤‴ (≮⇒≥ q≮2+k))
+          $ λ {d} d<k d-prime → k-rough-n d<k d-prime ∘ flip ∣-trans (quotient∣n k∣n)
+
+      prop : (2 + k) * product (factors res) ≡ 2 + n
+      prop = begin-equality
+        (2 + k) * product (factors res)
+          ≡⟨ cong ((2 + k) *_) (isFactorisation res) ⟩
+        (2 + k) * q
+          ≡⟨ *-comm (2 + k) q ⟩
+        q * (2 + k)
+          ≡˘⟨ _∣_.equality k∣n ⟩
+        2 + n ∎
+
+------------------------------------------------------------------------
+-- Properties of a factorisation
+
+factorisation≥1 : ∀ {as} → All Prime as → product as ≥ 1
+factorisation≥1 {[]} [] = s≤s z≤n
+factorisation≥1 {suc a ∷ as} (pa ∷ asPrime) = *-mono-≤ {1} {1 + a} (s≤s z≤n) (factorisation≥1 asPrime)
+
+factorisationPullToFront : ∀ {as} {p} → Prime p → p ∣ product as → All Prime as → ∃[ as′ ] as ↭ (p ∷ as′)
+factorisationPullToFront {[]} {suc (suc p)} pPrime p∣Πas asPrime = contradiction (∣1⇒≡1 p∣Πas) λ ()
+factorisationPullToFront {a ∷ as} {p} pPrime p∣aΠas (aPrime ∷ asPrime)
+  with euclidsLemma a (product as) pPrime p∣aΠas
+... | inj₂ p∣Πas = Π.map (a ∷_) step ih
+  where
+    ih : ∃[ as′ ] as ↭ (p ∷ as′)
+    ih = factorisationPullToFront pPrime p∣Πas asPrime
+
+    step : ∀ {as′} → as ↭ p ∷ as′ → a ∷ as ↭ p ∷ a ∷ as′
+    step {as′} as↭p∷as′ = begin
+      a ∷ as      ↭⟨ prep a as↭p∷as′ ⟩
+      a ∷ p ∷ as′ ↭⟨ swap a p refl ⟩
+      p ∷ a ∷ as′ ∎
+      where open PermutationReasoning
+
+... | inj₁ p∣a with ∣p⇒≡1∨≡p p aPrime p∣a
+...   | inj₁ refl = ⊥-elim pPrime
+...   | inj₂ refl = as , ↭-refl
+
+factorisationUnique′ : (as bs : List ℕ) → product as ≡ product bs → All Prime as → All Prime bs → as ↭ bs
+factorisationUnique′ [] [] Πas≡Πbs asPrime bsPrime = refl
```

Comment:
This lemma would simplify a lot if you made your `PrimeFactorisation` a *sorted* list of naturals.

Reply:
I didn't want to make the lists sorted, because I wanted to leave things open for other factorisation algorithms, which don't necessarily naturally make a sorted list. And I'm not at all convinced that it would make things simpler enough for it to be worth it